### ISSUE TITLE
Formatter processed on a per file basis

### ISSUE THREAD

**Issue by lbesson** [palantir/tslint#656]
As far as I understand, rules are run on a per file basis. This means that formatter is called once for each file.

This is a problem if for example we want to output checkstyle file, since the xml and checkstyle tag must be added by other means.

Else, we would get output such as

```xml
<?xml version="1.0" encoding="utf-8"?>
<checkstyle version="4.3">
  <file name="file1.ts">
    <error line="1" "column="1" "severity="error" "blah" />
    <error line="2" "column="1" "severity="error" "bluh" />
  </file>
</checkstyle>
<checkstyle version="4.3">
  <file name="file2.ts">
    <error line="1" "column="1" "severity="error" "blah" />
    <error line="2" "column="1" "severity="error" "bluh" />
  </file>
</checkstyle>
```

which is incorrect

**Comment by lbesson** [comment]
Looks similar to https://github.com/palantir/tslint/issues/379

**Comment by lbesson** [comment]
Should probably be solved once https://github.com/palantir/tslint/issues/680 is ok?

**Comment by Sweetchuck** [comment]
I wrote a little npm package which implements 2 new formatters for _tslint_, `yaml` and `checkstyle`.
As @lbesson pointed out that there is a problem with the checkstyle format because the formatter called with one file at a time and all the output saved into one file (or stdout). This method leads to an invalid XML file.

Luckily this method allows to generate a syntactically correct multi-document YAML output and it can be converted to any format including a correct checkstyle XML file.

This npm package also can be a temporary solution for #250 (Suggestion: Add checkstyle formatter)

https://github.com/Sweetchuck/npm-tslint-formatters

**Comment by JKillian** [comment]
#379 contains some new discussion on this issue - discussion can continue here, this is something that should get fixed!

**Comment by EarthCitizen** [comment]
When I originally looked at this and commented in #379, I was not paying attention to the fact that the Linter constructor wants the file name AND the contents of the file. This makes refactoring class Linter a bit difficult. If there needs to be backwards compatibility with a public API, then perhaps mark that class deprecated and start over in a new class.

If that were to happen, the new class should get the file contents itself when it is linting and not receive them in the constructor.

**Comment by caugner** [comment]
I can see two solutions for this issue:

1. Let `#processFile` return the results per file without calling the formatter (or writing to the outputStream respectively), collect all results and hand it to the formatter (and write it to the outputstream) only once at the end of the tslint run.
2. Reuse the formatter instance across one tslint run so that it can maintain state about whether (a) previous failures have been reported and/or (b) a header has been written already. Maybe it would help even more to add formatter methods `#formatHeader` and `#formatFooter` and call these exactly once before processing.

Unfortunately in either case existing (custom) formatters will need to be adapted, even though I would hope that we can simply add empty default implementations in the [AbstractFormatter](https://github.com/palantir/tslint/blob/2cfdd33bb49cf9d38c6538992812cbb746a25f87/src/language/formatter/abstractFormatter.ts).

**Comment by JKillian** [comment]
@caugner: I feel like your first proposal, or some variation of it makes sense. I think at least some of the formatters would be able to work as is, which is an added benefit.
